google-cloud-pubsublite/src/test/java/com/google/cloud/pubsublite/internal/BlockingPullSubscriberImplTest.java

```diff
@@ -0,0 +1,233 @@
+        SeekRequest.newBuilder()
+            .setCursor(Cursor.newBuilder().setOffset(initialOffset.value()).build())
+            .build();
+    when(underlying.seek(seek)).thenReturn(ApiFutures.immediateFuture(initialOffset));
+    FlowControlRequest flow =
+        FlowControlRequest.newBuilder()
+            .setAllowedBytes(flowControlSettings.bytesOutstanding())
+            .setAllowedMessages(flowControlSettings.messagesOutstanding())
+            .build();
+    when(underlyingFactory.newSubscriber(any()))
+        .thenAnswer(
+            args -> {
+              messageConsumer = args.getArgument(0);
+              return underlying;
+            });
+    doAnswer(
+            (Answer<Void>)
+                args -> {
+                  errorListener = args.getArgument(0);
+                  return null;
+                })
+        .when(underlying)
+        .addListener(any(), any());
+
+    subscriber =
+        new BlockingPullSubscriberImpl(underlyingFactory, flowControlSettings, initialSeek);
+
+    InOrder inOrder = inOrder(underlyingFactory, underlying);
+    inOrder.verify(underlyingFactory).newSubscriber(any());
+    inOrder.verify(underlying).addListener(any(), any());
+    inOrder.verify(underlying).startAsync();
+    inOrder.verify(underlying).awaitRunning();
+    inOrder.verify(underlying).seek(seek);
+    inOrder.verify(underlying).allowFlow(flow);
+
+    assertThat(messageConsumer).isNotNull();
+    assertThat(errorListener).isNotNull();
+  }
+
+  @Test
+  public void closeStops() {
+    when(underlying.stopAsync()).thenReturn(underlying);
+    subscriber.close();
+    verify(underlying).stopAsync();
+    verify(underlying).awaitTerminated();
+  }
+
+  @Test
+  public void onDataAfterErrorThrows() {
+    CheckedApiException expected = new CheckedApiException(StatusCode.Code.INTERNAL);
+    errorListener.failed(null, expected);
+    ExecutionException e = assertThrows(ExecutionException.class, () -> subscriber.onData().get());
+    assertThat(expected).isEqualTo(e.getCause());
+  }
+
+  @Test
+  public void onDataBeforeErrorThrows() throws Exception {
+    CheckedApiException expected = new CheckedApiException(StatusCode.Code.INTERNAL);
+    Future<?> future =
+        executorService.submit(
+            () -> {
+              ExecutionException e =
+                  assertThrows(ExecutionException.class, () -> subscriber.onData().get());
+              assertThat(expected).isEqualTo(e.getCause());
+            });
+    Thread.sleep(1000);
+    assertThat(future.isDone()).isFalse();
+
+    errorListener.failed(null, expected);
+    future.get();
+  }
+
+  @Test
+  public void onDataSuccess() throws Exception {
+    SequencedMessage message =
+        SequencedMessage.of(Message.builder().build(), Timestamps.EPOCH, Offset.of(12), 30);
+    Future<?> future = executorService.submit(() -> subscriber.onData().get());
+    messageConsumer.accept(ImmutableList.of(message));
+    future.get();
+  }
+
+  @Test
```

Comment:
this is not the correct behavior. They should just alias the same future.

Reply:
This is removed.

Reply:
also note, java is pass by value, so it's a copy of future1.